### PR TITLE
Resolves coredump caused by `tf.data.experimental.save` with prefetch

### PR DESCRIPTION
**Pull request by ashahab**
Repeat and prefetch in combination cause the snapshot reader Initialize function to be invoked multiple times.
However, there is nothing to prefetch on the very last iteration. This results in Prefetch issuing a CancelThreads call while the snapshot thread is trying to initialize. See https://github.com/tensorflow/tensorflow/blob/6446dda92eaadf11d22377e2354307642d739d73/tensorflow/core/kernels/data/prefetch_dataset_op.cc#L151

Currently the dataset reference counting is done asymmetrically. The reference increment happens at the end of initialization, where as the reference decrement
happens in a destructor. When prefetch cancels the snapshot thread, it errors out of the initialization function. And stops calling the reference increment. However, the reference decrement happens regardless, as it is in the destructor which always is invoked during cleanup. This results in an attempt to decrement the null dataset pointer, and therefore a segmentation fault.
This is different from all other dataset ops, where the dataset reference increment happens in the constructor and the decrement happens in the destructor, which are symmetric.

The solution to this is to ensure that the dataset reference is always initialized to nullptr, and to check for null when decrementing the dataset reference.

### CODE REVIEW

File: tensorflow/core/kernels/data/experimental/io_ops.cc [hunk 1]
```diff
@@ -262,7 +262,12 @@ class LoadDatasetOp::Dataset : public DatasetBase {
     explicit Iterator(const Params& params)
         : DatasetIterator<Dataset>(params) {}
 
-    ~Iterator() override { input_->Unref(); }
+    ~Iterator() override {
+      // Ensure not cleaning up uninitialized input_
+      if(input_) {
+        input_->Unref();
```
Comment (yangustc07) [inline tensorflow/core/kernels/data/experimental/io_ops.cc:268]: Thanks for the change. Do you know why we need this? I don't see this class calls `Ref()`. There is a `Unref` call in the base iterator.
https://github.com/tensorflow/tensorflow/blob/a9cf3a0e4b419630f0183b0cc4e48e0641a62721/tensorflow/core/framework/dataset.cc#L701

Comment (ashahab) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: @yangustc07  thank you for the review.
This is the corresponding call to Ref(): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/io_ops.cc#L338

Comment (ashahab) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: And this Ref() and the input_ initialization does not happen because of a cancelled prefetch:
```bash
tensorflow/core/kernels/data/experimental/io_ops.cc:282] Status Cancelled:
        For thread id 1397804166 thread name: tf_data_iterator_resource/1397804166*** Begin stack trace ***
        tensorflow::CurrentStackTrace()
        tensorflow::data::experimental::LoadDatasetOp::Dataset::Iterator::Initialize(tensorflow::data::IteratorContext*)
        tensorflow::data::DatasetBase::MakeIterator(tensorflow::data::IteratorContext*, tensorflow::data::IteratorBase const*, std::string const&, std::unique_ptr<tensorflow::data::IteratorBase, std::default_delete<tensorflow::data::IteratorBase> >*) const
        tensorflow::data::ShuffleDatasetOpBase::ShuffleDatasetBase::Iterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::BatchDatasetOp::Dataset::Iterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::RepeatDatasetOp::Dataset::ForeverIterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::PrefetchDatasetOp::Dataset::Iterator::PrefetchThread(std::shared_ptr<tensorflow::data::IteratorContext> const&)

        std::_Function_handler<void (), std::_Bind<void (*(std::function<void ()>, std::shared_ptr<tensorflow::Notification>))(std::function<void ()> const&, std::shared_ptr<tensorflow::Notification>)> >::_M_invoke(std::_Any_data const&)
        tensorflow::UnboundedWorkQueue::PooledThreadFunc()
```

Comment (yangustc07) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: I see, it seems this `Ref` is not needed for the same reason mentioned in https://github.com/tensorflow/tensorflow/commit/858a5698a7b4ee95befa2e9c3d7aaa0a8170ec54. What do you think?

Comment (ashahab) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: Actually removing Ref results in a segfault without a stracktrace in this case. I'm still investigating that.

Comment (ashahab) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: @yangustc07 Did you have any other feedback?

File: tensorflow/python/data/experimental/kernel_tests/io_test.py [hunk 2]
```diff
@@ -111,6 +111,19 @@ def testOptionalElementSpec(self):
...
+    dataset1 = dataset_ops.Dataset.from_tensor_slices(np.random.rand(16, 32))
+    io.save(dataset1, self._test_dir)
+    dataset = io.load(self._test_dir)
+    dataset = dataset.shuffle(buffer_size=16)
+    dataset = dataset.batch(16)
+    dataset = dataset.repeat()
+    dataset = dataset.prefetch(1)
+    next_element = self.getNext(dataset)
+    for _ in range(30):
+      self.evaluate(next_element())
```
Comment (yangustc07) [inline tensorflow/python/data/experimental/kernel_tests/io_test.py:127]: Can you leave 2 blank lines here? Similar to the original code.

File: tensorflow/python/data/experimental/kernel_tests/io_test.py [hunk 3]
```diff
@@ -111,6 +111,19 @@ def testOptionalElementSpec(self):
     dataset_loaded = io.load(self._test_dir)
     self.assertDatasetsEqual(dataset, dataset_loaded)
 
+  @combinations.generate(test_base.eager_only_combinations())
+  def testRepeatAndPrefetch(self):
+    """This test reproduces github.com/tensorflow/tensorflow/issues/49165"""
+    dataset1 = dataset_ops.Dataset.from_tensor_slices(np.random.rand(16, 32))
```
Comment (yangustc07) [inline tensorflow/python/data/experimental/kernel_tests/io_test.py:117]: I think you need to import `np`.

File: tensorflow/core/kernels/data/experimental/io_ops.cc (hunk 1 above)
Comment (yangustc07) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: I'm trying this test on my machine. I can produce the same segfault. I'll update you soon. I'm sorry for the inconvenience.

Comment (yangustc07) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:268]: Yes, I agree with you. `Ref()` is needed here because `reader_output` is destroyed in `InitializeInput`.

Thanks for the fix!

File: tensorflow/core/kernels/data/experimental/io_ops.cc [hunk 4]
```diff
@@ -262,7 +262,12 @@ class LoadDatasetOp::Dataset : public DatasetBase {
     explicit Iterator(const Params& params)
         : DatasetIterator<Dataset>(params) {}
 
-    ~Iterator() override { input_->Unref(); }
+    ~Iterator() override {
+      // Ensure not cleaning up uninitialized input_
```
Comment (yangustc07) [inline tensorflow/core/kernels/data/experimental/io_ops.cc:266]: The code is self-explanatory. Maybe remove the comment? Thank you.

File: tensorflow/tools/pip_package/setup.py [hunk 5]
```diff
@@ -50,7 +50,7 @@
 # result for pip.
 # Also update tensorflow/tensorflow.bzl and
 # tensorflow/core/public/version.h
-_VERSION = '2.6.0'
+_VERSION = '2.6.0.'
```
Comment (yangustc07) [inline tensorflow/tools/pip_package/setup.py:53]: Is this needed? It looks not related to the fix.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 6]
```diff
@@ -78,6 +78,50 @@ void TrtShapeOptimizationProfile::ImplicitBatchModeCompatibleStrategy(
   }
```
Comment (yangustc07) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc:78]: Did you intend to make this change in another PR?

Comment (ashahab) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: removed

File: tensorflow/core/kernels/data/experimental/io_ops.cc [hunk 7]
```diff
@@ -262,7 +262,11 @@ class LoadDatasetOp::Dataset : public DatasetBase {
     explicit Iterator(const Params& params)
         : DatasetIterator<Dataset>(params) {}
 
-    ~Iterator() override { input_->Unref(); }
+    ~Iterator() override {
+      if(input_) {
```
Comment (yangustc07) [inline tensorflow/core/kernels/data/experimental/io_ops.cc:266]: To keep the style consistent, could you add a space after `if`?

**Review by yangustc07 — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/data/experimental/io_ops.cc (hunk 7 above)
Comment (yangustc07) [reply tensorflow/core/kernels/data/experimental/io_ops.cc:266]: Sorry, I meant to approve, not requesting change.

**Review by yangustc07 — APPROVED** [review]

**Review by yangustc07 — APPROVED** [review]

**Comment by ashahab** [comment]
Can you merge this? I plan to cherry-pick to 2.4 and 2.5 soon.

**Comment by yangustc07** [comment]
I have approved your change internally. It requires approval from another reviewer.

**Comment by ashahab** [comment]
Can we request @jsimsa for a review? He was helping me with another PR.

**Comment by yangustc07** [comment]
Your change has been approved. It will be submitted soon. Currently, an unrelated already-failing test blocks the submission. It will be retried.
